pkg/autodiscovery/listeners/snmp.go

```diff
@@ -0,0 +1,351 @@
+		return
+	}
+	for _, deviceIP := range devices {
+		entityID := config.Digest(deviceIP.String())
+		l.createService(deviceIP.String(), adIdentifier, entityID)
+	}
+}
+
+func writeCache(cacheKey string, devices []string) {
+	cacheValue, err := json.Marshal(devices)
+	if err != nil {
+		log.Errorf("Couldn't marshal cache: %s", err)
+		return
+	}
+	err = persistentcache.Write(cacheKey, string(cacheValue))
+	if err != nil {
+		log.Errorf("Couldn't write cache: %s", err)
+	}
+}
+
+// Don't make it a method, to be overridden in tests
+var worker = func(l *SNMPListener, jobs <-chan snmpSubnet) {
+	for {
+		select {
+		case <-l.stop:
+			log.Debug("Stopping SNMP worker")
+			return
+		case subnet := <-jobs:
+			log.Debugf("Handling IP %s", subnet.currentIP.String())
+			l.checkDevice(subnet.adIdentifier, subnet.currentIP.String(), subnet.config, subnet.defaultParams)
+		default:
+		}
+	}
+}
+
+func (l *SNMPListener) checkDevice(adIdentifier string, deviceIP string, config util.SNMPConfig, defaultParams *gosnmp.GoSNMP) {
+	params := *defaultParams
+	params.Target = deviceIP
+	entityID := config.Digest(deviceIP)
+	err := params.Connect()
+	if err != nil {
+		log.Debugf("SNMP connect to %s error: %v", deviceIP, err)
+		l.deleteService(entityID)
+	} else {
+		defer params.Conn.Close()
+
+		oids := []string{"1.3.6.1.2.1.1.2.0"}
+		v, err := params.Get(oids)
+		if err != nil || len(v.Variables) < 1 || v.Variables[0].Value == nil {
+			log.Debugf("SNMP get to %s error: %v", deviceIP, err)
+			l.deleteService(entityID)
+		} else {
+			log.Debugf("SNMP get to %s success: %v", deviceIP, v.Variables[0].Value)
+			l.createService(deviceIP, adIdentifier, entityID)
+		}
+	}
+}
+
+func (l *SNMPListener) checkDevices() {
+	subnets := []snmpSubnet{}
+	for i, config := range l.config.Configs {
+		ipAddr, ipNet, err := net.ParseCIDR(config.Network)
+		if err != nil {
+			return
+		}
+
+		defaultParams, err := config.BuildSNMPParams()
+		if err != nil {
+			log.Error(err)
+			continue
+		}
+		adIdentifier := fmt.Sprintf("snmp_%d", i)
+
+		startingIP := ipAddr.Mask(ipNet.Mask)
+		currentIP := make(net.IP, len(startingIP))
+		copy(currentIP, startingIP)
+
+		configHash := config.Digest(config.Network)
+		cacheKey := fmt.Sprintf("snmp:%s", configHash)
+		l.loadCache(config, adIdentifier, cacheKey)
+
+		subnet := snmpSubnet{
+			adIdentifier:  adIdentifier,
+			config:        config,
+			defaultParams: defaultParams,
+			startingIP:    startingIP,
+			currentIP:     currentIP,
+			network:       *ipNet,
+			cacheKey:      cacheKey,
+		}
+		subnets = append(subnets, subnet)
+	}
+
+	if l.config.Workers == 0 {
+		l.config.Workers = 2
+	}
+
+	if l.config.DiscoveryInterval == 0 {
+		l.config.DiscoveryInterval = 3600
+	}
+
+	jobs := make(chan snmpSubnet)
+	for w := 0; w < l.config.Workers; w++ {
+		go worker(l, jobs)
+	}
+
+	subnetsDone := 0
+	started := time.Now()
+	discoveryInterval := time.Duration(l.config.DiscoveryInterval) * time.Second
```

Comment:
I think we can use a [ticker](https://golang.org/pkg/time/#Ticker) here

Reply:
Done!